---
title: "Tool popularity analysis"
output: html_document
---

# What is this?
This is a analysis of bioconda tool download metrics in order to prioritize our efforts in tool descriptor creation on Truwl

```{r}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(dplyr)
library(stringr)
```

```{r}
dl<-read.table("2020.clean.txt",col.names=c("package","downloads"))

getSummary<-function(recipe){
  #cat(recipe,"\n")
  if(file.exists(paste0("../bioconda-recipes/recipes/",recipe,"/meta.yaml"))){
    metayaml<-tryCatch({yaml.load_file(paste0("../bioconda-recipes/recipes/",recipe,"/meta.yaml"),eval.expr=FALSE)},
                       error=function(cond){return(NULL)})
    if(is.null(metayaml)){return("")}
    if(exists('about', where=metayaml)){
      #cat("about")
      about<-metayaml$about
      if(exists('summary', where=about)){
        return(metayaml$about$summary)
      }
    }
  }
  return("")
}

getBiotools<-function(tool){
  cat(tool,"\n")
  tool<-str_replace(string=tool,pattern = '^bioconductor-',replacement = '')
  tool<-str_replace(string=tool,pattern = '^perl-',replacement = '')
  terms<-tryCatch({
  jsonlite::fromJSON(paste0("https://bio.tools/api/tool/",tool))$topic$term},error=function(cond){return(NULL)})
  if(is.null(terms)){return("None")}
  else{
    return(paste(terms,sep=", ",collapse=","))
    #return(data.frame(package=tool,term=terms))
  }
}
```

In 2020, there were `r nrow(dl)` tools in bioconda. `r dl %>% filter(str_detect(package,regex('(bioconductor|perl)'))) %>% summarize(cnt=n_distinct(package)) %>% pull ` of those were Bioconductor packages or Perl modules.

```{r}
dl %>% rowwise() %>% dplyr::mutate(biotools=getBiotools(package)) -> btdl #slow
btdl %>% dplyr::mutate(summary=getSummary(package)) -> annodl
```

```{r}
catannodl <- annodl %>% rowwise() %>% dplyr::mutate(bioconductor=str_detect(package,'bioconductor')) %>%
  dplyr::mutate(perl=str_detect(package,'perl')) %>%
  dplyr::mutate(rna=str_detect(summary,fixed('rna', ignore_case=TRUE))) %>%
  dplyr::mutate(alignment=str_detect(summary,regex('(alignment|BAM)', ignore_case=TRUE))) %>%
  dplyr::mutate(assembly=str_detect(summary,fixed('assemb', ignore_case=TRUE))) %>%
  dplyr::mutate(affy=str_detect(summary,fixed('affy', ignore_case=TRUE))) %>%
  dplyr::mutate(metagnomics=str_detect(summary,fixed('metageno', ignore_case=TRUE))) %>%
  dplyr::mutate(proteomics=str_detect(summary,regex('(amino|prote)', ignore_case=TRUE))) %>%
  dplyr::mutate(variation=str_detect(summary,regex('(VCF|variant|SNP|GATK)', ignore_case=TRUE))) %>%
  dplyr::mutate(sequence=str_detect(summary,regex('(fasta|fastq|reads|sequencing|adapter)', ignore_case=TRUE))) %>%
  dplyr::mutate(seqfeatures=str_detect(summary,regex('(motif|k-mer|repeat)', ignore_case=TRUE))) %>%
  dplyr::mutate(genomefeatures=str_detect(summary,regex('(bed|feature)', ignore_case=TRUE))) %>%
  dplyr::mutate(chipseq=str_detect(summary,regex('(ChIP-Seq)', ignore_case=TRUE))) %>%
  dplyr::mutate(nanopore=str_detect(summary,regex('(Nanopore)', ignore_case=TRUE))) %>%
  dplyr::mutate(cancer=str_detect(summary,regex('(onco|cancer|somatic)', ignore_case=TRUE))) %>%
  dplyr::mutate(population=str_detect(summary,regex('(population|GWAS)', ignore_case=TRUE))) %>%
  dplyr::mutate(vizualization=str_detect(summary,regex('(vizual|visual)', ignore_case=TRUE))) %>% 
  dplyr::mutate(micro=str_detect(summary,regex('(micro|pathogen|strain|viral)', ignore_case=TRUE))) %>% 
  dplyr::mutate(phylo=str_detect(summary,regex('(phylo)', ignore_case=TRUE))) %>%
  dplyr::mutate(cnv=str_detect(summary,regex('(CNV|SV)', ignore_case=TRUE))) 

uncategorized <- catannodl %>% 
  filter_at(vars(-package, -downloads, -summary), all_vars(isFALSE(.))) 

```

```{r}
catannodl %>% filter(bioconductor==FALSE && perl==FALSE) %>% arrange(-downloads) %>% select(package,downloads,summary) %>% head(100) %>% knitr::kable()
catannodl %>% filter(bioconductor==FALSE && perl==FALSE) %>% arrange(-downloads) %>% select(package,downloads,summary) %>% head(100) -> top100
write.table(top100,file="top100.tsv",sep="\t")
```

```{r}
catannodl %>% arrange(-downloads) %>% select(-package,-downloads,-summary) %>% colSums()
```